internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,365 @@
+		}
+	}
+
+	if !rancherCluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		return reconcileDelete(ctx, capiCluster)
+	}
+
+	return r.reconcileNormal(ctx, capiCluster, rancherCluster)
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcileNormal(ctx context.Context, capiCluster *clusterv1.Cluster,
+	rancherCluster *managementv3.Cluster,
+) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	err := r.RancherClient.Get(ctx, client.ObjectKeyFromObject(rancherCluster), rancherCluster)
+	if apierrors.IsNotFound(err) {
+		shouldImport, err := util.ShouldAutoImport(ctx, log, r.Client, capiCluster, importLabelName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if !shouldImport {
+			log.Info("not auto importing cluster as namespace or cluster isn't marked auto import")
+			return ctrl.Result{}, nil
+		}
+
+		if err := r.RancherClient.Create(ctx, &managementv3.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    capiCluster.Namespace,
+				GenerateName: "c-",
+				Labels: map[string]string{
+					capiClusterOwner:          capiCluster.Name,
+					capiClusterOwnerNamespace: capiCluster.Namespace,
+					ownedLabelName:            "",
+				},
+			},
+			Spec: managementv3.ClusterSpec{
+				DisplayName: capiCluster.Name,
+				Description: "CAPI cluster imported to Rancher",
+			},
+		}); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error creating rancher cluster: %w", err)
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to fetch rancher cluster %s", client.ObjectKeyFromObject(rancherCluster)))
+
+		return ctrl.Result{}, err
+	}
+
+	if managementv3.ClusterConditionAgentDeployed.IsTrue(rancherCluster) {
+		log.Info("agent already deployed, no action needed")
+		return ctrl.Result{}, nil
+	}
+
+	// get the registration manifest
+	manifest, err := getClusterRegistrationManifest(ctx, rancherCluster.Name, rancherCluster.Name, r.RancherClient, r.InsecureSkipVerify)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if manifest == "" {
+		log.Info("Import manifest URL not set yet, requeue")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	log.Info("Creating import manifest")
+
+	remoteClient, err := r.remoteClientGetter(ctx, capiCluster.Name, r.Client, client.ObjectKeyFromObject(capiCluster))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting remote cluster client: %w", err)
+	}
+
+	if err := createImportManifest(ctx, remoteClient, strings.NewReader(manifest)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating import manifest: %w", err)
+	}
+
+	log.Info("Successfully applied import manifest")
+
+	return ctrl.Result{}, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) rancherClusterToCapiCluster(ctx context.Context, clusterPredicate predicate.Funcs) handler.MapFunc {
+	log := log.FromContext(ctx)
+
+	return func(_ context.Context, o client.Object) []ctrl.Request {
+		cluster, ok := o.(*managementv3.Cluster)
```

Comment:
This code previously didn't have any static assumption on the source object for the events, so there was no dependency on v1.Cluster. Now it has a v3.Cluster check here I think is also not needed. This is all driven by the controller watches setup, where you specify what cluster to watch. I the v1 vs v3 difference in labels could be a simple if case with additional logic here.

Reply:
Its the normal convention to check the type and return nil if you can't cast it:

- https://github.com/rancher-sandbox/rancher-turtles/blob/b3ea1a3a63c448f40b0b2f13015ca586fff1b193/internal/controllers/import_controller.go#L383
- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controlplane/eks/controllers/awsmanagedcontrolplane_controller.go#L434

The fact we didn't have it previously is a mistake i would say.

Reply:
This check can also be found in upstream projects, for example: https://github.com/kubernetes-sigs/cluster-api/blob/0045baf02ce542fad0f96e42ab2decc41822256e/controlplane/kubeadm/internal/controllers/controller.go#L576

Reply:
This was a convention when controller-runtime may receive a partial object from an API server which will not be valid by schema for deserialization. This is not happening now. As we only need a namespace/name here, and we basically rely on conformity with client.Object interface, I don’t think all conventions should be preserved. We don’t rely on any api specific fields from v1 cluster or v3 cluster here. You can quote me on that, but I don’t see harm in dropping this anywhere now for the new projects.

Reply:
That's a good point, if there has been a change in controller-runtime that makes this check redundant then yes for sure we should update the conventions we use/follow.

@Danil-Grigorev - out of curiosity so you know when the change was introduced into CR?

Reply:
Thanks @Danil-Grigorev. Considering this, I removed the type check.

Reply:
Thanks @Danil-Grigorev. Considering this, I removed the type check.

Reply:
@richardcase [Here](https://github.com/kubernetes-sigs/controller-runtime/blob/56159419231e985c091ef3e7a8a3dee40ddf1d73/pkg/reconcile/reconcile.go#L137) is for example a conversion without error check and it is a generic one every controller will now use. Additionally here we only need namespace/name, so for the sake of this method only this info required. And as we have it in available, then it was typechecked before.